File: imacals-web/src/services/wishlist.ts
import { api } from '@/services/api';
import type { Product } from '@/services/catalog';

// A wishlist as returned by the index endpoint — basic fields plus the live item count.
export interface WishlistSummary {
  id: string;
  organization_id: string;
  customer_id: string;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string;
  item_count: number;
}

// A single wishlist item with the joined product snapshot — the shape the storefront renders.
export interface WishlistItem {
  id: string;
  wishlist_id: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
  product: Product;
}

// A wishlist detail — basic fields flattened together with its items.
export interface WishlistDetail {
  id: string;
  organization_id: string;
  customer_id: string;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string;
  items: WishlistItem[];
}

export const wishlistService = {
  list: (): Promise<WishlistSummary[]> =>
    api.get<WishlistSummary[]>('/wishlists'),

  show: (id: string): Promise<WishlistDetail> =>
    api.get<WishlistDetail>(`/wishlists/${id}`),

  create: (name: string, description?: string): Promise<WishlistSummary> =>
    api.post<WishlistSummary>('/wishlists', {
      name,
      description: description ?? null,
    }),

  update: (id: string, payload: { name?: string; description?: string | null }): Promise<WishlistSummary> =>
    api.put<WishlistSummary>(`/wishlists/${id}`, payload),

  remove: (id: string): Promise<void> =>
    api.delete<void>(`/wishlists/${id}`),

  addItem: (wishlistId: string, productId: string, notes?: string): Promise<WishlistDetail> =>
    api.post<WishlistDetail>(`/wishlists/${wishlistId}/items`, {
      product_id: productId,
      notes: notes ?? null,
    }),

  removeItem: (wishlistId: string, itemId: string): Promise<void> =>
    api.delete<void>(`/wishlists/${wishlistId}/items/${itemId}`),
};

File: imacals-api/src/models/wishlist.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

use super::product::CatalogProduct;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Wishlist {
    pub id: Uuid,
    pub organization_id: Uuid,
    pub customer_id: Uuid,
    pub name: String,
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deleted_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WishlistItem {
    pub id: Uuid,
    pub wishlist_id: Uuid,
    pub product_id: Uuid,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deleted_at: Option<DateTime<Utc>>,
}

// Wishlist item with its product snapshot embedded — what the storefront renders as a card.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WishlistItemWithProduct {
    pub id: Uuid,
    pub wishlist_id: Uuid,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub product: CatalogProduct,
}

// Summary used in list views — wishlist fields flattened together with the current item count.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WishlistSummary {
    #[serde(flatten)]
    pub wishlist: Wishlist,
    pub item_count: i64,
}

// Detail used in show views — wishlist fields flattened together with the active items.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WishlistDetail {
    #[serde(flatten)]
    pub wishlist: Wishlist,
    pub items: Vec<WishlistItemWithProduct>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct CreateWishlistSchema {
    // Optional: when omitted, the service resolves the customer from the acting user.
    pub customer_id: Option<Uuid>,
    #[validate(length(min = 1, max = 255, message = "Wishlist name must be between 1 and 255 characters"))]
    pub name: String,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct UpdateWishlistSchema {
    #[validate(length(min = 1, max = 255, message = "Wishlist name must be between 1 and 255 characters"))]
    pub name: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct AddWishlistItemSchema {
    pub product_id: Uuid,
    pub notes: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn create_wishlist_schema_requires_name() {
        let result: Result<CreateWishlistSchema, _> = serde_json::from_str("{}");
        assert!(result.is_err());
    }

    #[test]
    fn create_wishlist_schema_accepts_valid_payload() {
        let json = r#"{"name": "Reorder list", "description": "For next visit"}"#;
        let schema: CreateWishlistSchema = serde_json::from_str(json).unwrap();
        assert!(schema.validate().is_ok());
    }

    #[test]
    fn add_item_requires_product_id() {
        let result: Result<AddWishlistItemSchema, _> = serde_json::from_str("{}");
        assert!(result.is_err());
    }
}
